src/middlewares/Cache.spec.ts

```diff
@@ -0,0 +1,128 @@
+import {Cache, ICacheMiddlewareStore} from "./Cache";
+import {FetchResponse} from "./Fetch";
+
+// noinspection TsLint
+declare const Response: FetchResponse<string>;
+declare const localStorage: ICacheMiddlewareStore;
+
+describe("Cache", () => {
+    it("is defined", () => {
+        expect(Cache).toBeDefined();
+    });
+
+    describe("constructor()", () => {
+        it("can be instantiated", () => {
+            expect(new Cache() instanceof Cache).toBeTruthy();
+        });
+
+        it("accepts storage engine in instantiation", () => {
+            expect(new Cache(localStorage) instanceof Cache).toBeTruthy();
+        });
+    });
+
+    describe("process()", () => {
+        it("is a function", () => {
+            expect(typeof (new Cache()).process).toBe("function");
+        });
+
+        it("calls next() if cache not enabled", (done) => {
+            const cacheMw: Cache = new Cache();
+            const mock: Promise<any> = Promise.resolve(new Response("fetched response", undefined));
+            const cached = cacheMw.process({ url: "/test" }, () => {
+                setTimeout(() => done(), 20);
+
+                return mock;
+            });
+
+            expect(cached instanceof Promise).toBeTruthy();
+        });
+
+        it("calls next() if cache expired", (done) => {
+            const cache = { enable: true, maxAge: 10 }; // 10 seconds
+            const mock: Promise<any> = Promise.resolve(new Response("fetched response", undefined));
+
+            // Seed the cache with expired timestamp for test
+            localStorage.setItem(
+                "/test",
+                JSON.stringify({
+                    value: "cached response",
+                    type: "basic",
+                    url: "/test",
+                    status: 200,
+                    statusText: "OK",
+                    timestamp: +Date.now() - 12000, // 12 seconds before
+                    headers: { "Content-Type": "text/html" }
+                })
+            );
+
+            const cached = new Cache(localStorage).process({ url: "/test", cache }, () => {
+                setTimeout(() => done(), 20);
+
+                return mock;
+            });
+
+            expect(cached instanceof Promise).toBeTruthy();
+        });
+
+        it("calls next() if cache invalid/corrupt", (done) => {
+            const cache = { enable: true, maxAge: 10 };
+            const mock: Promise<any> = Promise.resolve(new Response("fetched response", undefined));
+
+            // Seed the cache with invalid value
+            localStorage.setItem("/test1", "\"");
+
+            const cached = new Cache(localStorage).process({ url: "/test1", cache }, () => {
+                setTimeout(() => done(), 20);
+
+                return mock;
+            });
+
+            expect(cached instanceof Promise).toBeTruthy();
```

Comment:
the done() is called in the `next` callback because it verifies that `next()` has been actually called.
if we place `done()` somewhere else it doesnt guarantee if `next()` was called